src/Microsoft.AspNet.Server.Kestrel/Http/Frame.Templated.cs

```diff
@@ -0,0 +1,128 @@
+                     IPEndPoint localEndPoint)
+            : base(context, remoteEndPoint, localEndPoint)
+        {
+            _application = application;
+        }
+
+        public override async Task RequestProcessingAsync()
+        {
+            try
+            {
+                var terminated = false;
+                while (!terminated && !_requestProcessingStopping)
+                {
+                    while (!terminated && !_requestProcessingStopping && !TakeStartLine(SocketInput))
+                    {
+                        terminated = SocketInput.RemoteIntakeFin;
+                        if (!terminated)
+                        {
+                            await SocketInput;
+                        }
+                    }
+
+                    while (!terminated && !_requestProcessingStopping && !TakeMessageHeaders(SocketInput, _requestHeaders))
+                    {
+                        terminated = SocketInput.RemoteIntakeFin;
+                        if (!terminated)
+                        {
+                            await SocketInput;
+                        }
+                    }
+
+                    if (!terminated && !_requestProcessingStopping)
+                    {
+                        MessageBody = MessageBody.For(HttpVersion, _requestHeaders, this);
+                        _keepAlive = MessageBody.RequestKeepAlive;
+                        RequestBody = new FrameRequestStream(MessageBody);
+                        ResponseBody = new FrameResponseStream(this);
+                        DuplexStream = new FrameDuplexStream(RequestBody, ResponseBody);
+
+                        var context = _application.CreateContext(this);
+                        try
+                        {
+                            await _application.ProcessRequestAsync(context).ConfigureAwait(false);
+                        }
+                        catch (Exception ex)
+                        {
+                            ReportApplicationError(ex);
+                            _application.DisposeContext(context, ex);
+                        }
+                        finally
+                        {
+                            // Trigger OnStarting if it hasn't been called yet and the app hasn't
+                            // already failed. If an OnStarting callback throws we can go through
+                            // our normal error handling in ProduceEnd.
+                            // https://github.com/aspnet/KestrelHttpServer/issues/43
+                            if (!_responseStarted && _applicationException == null)
+                            {
+                                await FireOnStarting();
+                            }
+
+                            await FireOnCompleted();
+
+                            if (_applicationException == null)
+                            {
+                                _application.DisposeContext(context, null);
```

Comment:
The context might never be disposed with this logic. Example case: _application.ProcessRequestAsync completes successfully, but FireOnStarting/Completed sets _applicationException.